graph/pull.go

```diff
@@ -27,10 +18,28 @@ type ImagePullConfig struct {
 	OutStream   io.Writer
 }
 
+type Puller interface {
+	// Pull tries to pull `canonicalImageName` with tag `tag`, from `endpoint` with the config specified by `imagePullConfig`.
+	// `name` is the image's name on the server-side
+	// `sf` is an object that allows to stream back to the client (to be removed)
+	// Pull returns an error if any, as well as a boolean that determines whether to retry Pull on the next configured endpoint.
+	Pull(canonicalImageName, name, tag string, endpoint registry.APIEndpoint, imagePullConfig *ImagePullConfig, sf *streamformatter.StreamFormatter) (fallback bool, err error)
```

Comment:
these are all problems i had to inherit from older code, this interface sucks and will be replaced. I did it so that people initially reviewing this could see that v1 and v2 code paths are separated.

I started refactoring all this, but if i pushed it now it just creates even more changes.